scipy/optimize/_minpack_py.py

```diff
@@ -964,12 +964,20 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
             xdata = xdata[..., ~has_nan]
             ydata = ydata[~has_nan]
 
+            # Also omit the corresponding entries from sigma
+            if sigma is not None:
+                sigma = np.asarray(sigma)
+                if sigma.ndim == 1:
+                    sigma = sigma[~has_nan]
+                elif sigma.ndim == 2:
+                    sigma = sigma[~has_nan, ~has_nan]
+
     # Determine type of sigma
     if sigma is not None:
         sigma = np.asarray(sigma)
 
         # if 1-D or a scalar, sigma are errors, define transform = 1/sigma
-        if sigma.size == 1 or sigma.shape == (ydata.size, ):
+        if sigma.size == 1 or sigma.shape == (ydata.size,):
```

Comment:
I see now that you were trying to match this style. I'm not sure it was done that way. To begin with. ISTM these could look at the number of dimensions, not the size. That is what the comment says, after all.

Reply:
Yeah, I understand. I think checking the `ndim` in the present case works. If, after the omission of NaNs, the sigma still differs in size from the `ydata`, the existing code will catch it and raise the appropriate error.

Reply:
On second thought, what happens when sigma differs in size with ydata even before the omission of NaNs? For example consider the case when `sigma.size = 80` and `ydata.size = 100`. In this case the `sigma = sigma[~has_nan]` will raise an IndexError. Thus I believe we need to check the shape of sigma and not just the ndim.